File: backtrack/combination_sum/combination_sum.rb
def dfs(nums, target, index, path, res)
  if target.zero?
    res.push(path)
    return
  end
  (index...nums.length).each do |i|
    nxt = target - nums[i]
    break if nxt < 0
    dfs(nums, nxt, i, path + [nums[i]], res)
  end
end

def solve(candidates, target)
  res = []
  candidates.sort!
  dfs(candidates, target, 0, [], res)
  res
end

a = [2, 3, 6, 7]
p solve(a, 7) # should be [[2, 2, 3], [7]]

File: backtrack/combination_sum/combination_sum.ex
defmodule Solution do
  def dfs(_, target, _, _) when target < 0 do
    [[]]
  end
  def dfs(_, target, _, path) when target == 0 do
    [path]
  end
  def dfs(nums, target, index, path) do
    Enum.slice(nums, index..-1)
    |> Enum.with_index
    |> Enum.flat_map(fn {num, idx} -> dfs(nums, target - num, idx, path ++ [num]) end)
    |> Enum.filter(fn x -> length(x) != 0 end)
  end

  def solve(candidates, target) do
    candidates = Enum.sort(candidates)
    dfs(candidates, target, 0, [])
  end
end


a = [2, 3, 6, 7]
IO.inspect Solution.solve(a, 7), charlists: :as_lists  # should be [[2, 2, 3], [7]]
